images/cfw/opt/x1plus/lib/python/x1plus/services/x1plusd/actions.py

```diff
@@ -0,0 +1,90 @@
+"""
+Actions trigger mechanism
+
+Many different things on the X1 system need to take actions.  For instance,
+the on-board buttons have long been customizable with fixed actions, but
+there has been some demand to have them do more interesting things.  The
+actions mechanism solves this by adding a command set that can affect
+various parts of the system.  Additionally, the G-code interpreter can now
+be interrupted and send messages in real-time synchronously with the G-code
+stream from a print (see x1plus.mcproto).
+
+The actions trigger mechanism operates on JSON; an action is either a list
+of actions, or a dictionary with exactly one key (representing the name of
+an action), mapping to a parameter of type depending on the action.  A list
+executes each subaction in sequence, waiting for action completion.  For
+instance, consider the following JSON objects, that would all be conceivably
+valid actions, if their respective keys were implemented:
+
+  { "buzzer": { "duration": 0.2 } }
+  
+  [ { "shutter": true }, { "delay": 1 }, { "gcode": "M400 W0" } ]
+  
+  { "led": { "animation": "green", "timeout": 5 } }
+  
+  [ { "buzzer": { "duration": 0.2 } },
+    { "display": { "message": "The nozzle is ready for cold pull.  Apply tension on the filament, then tap OK.", "wait_for_confirmation": true } },
+    { "gcode": "M400 W0" } ]
+
+At some point in the future, it probably would be reasonable to write a
+predicate mechanism, for actions such as `{"if": {"condition":
+PREDICATE_OBJECT, "then": ACTION}}`, or `{"wait_until": {"condition":
+PREDICATE_OBJECT}}`.  For now, actions run in a straight line.
+
+It would probably be ergonomic to write action scripts in YAML and convert
+them to JSON before embedding them in G-code.
+"""
+
+import asyncio
+import logging
+
+logger = logging.getLogger(__name__)
+
+class ActionHandler():
+    ACTIONS = {}
+    
+    def __init__(self, daemon):
+        self.daemon = daemon
+    
```

Comment:
I like this a lot.  I'm also going to refactor `mcproto` in terms of this.

Reply:
I think that more idiomatic is to register with the subsystem ("register as one of the actions available on the system"), rather than with ActionHandler, whose only state is which x1plusd daemon it is associated with, so I hoisted that out into a `actions.register_handler` decorator.